### PR TITLE
tts: init at unstable-2020-06-17

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change

Mozilla TTS is a text to speech engine based on pytorch and tensorflow. @Mic92 and I have been successfully using it for the last few days in combination with https://rhasspy.readthedocs.io, which we might eventually package, not sure yet.

Demo: https://shells.darmstadt.ccc.de/~hexa/nixpkgs.wav

To get this running one needs to download a TTS and Vocoder model, check out https://github.com/mozilla/TTS/wiki/Released-Models, from where we pulled a commit, that matched a given version.

```
  systemd.services.tts = {
    after = [ "network.target" ];
    wantedBy = [ "multi-user.target" ];
    script = ''
      ${pkgs.tts}/bin/tts-server \
        --vocoder_config ./vocoder/config.json \
        --vocoder_checkpoint ./vocoder/checkpoint_1450000.pth.tar \
        --tts_config ./tts/config.json \
        --tts_checkpoint ./tts/checkpoint_130000.pth.tar
    '';
    serviceConfig = {
      User = "youruser";
      # needed for pulseaudio
      Environment = "XDG_RUNTIME_DIR=/run/user/1000";
      WorkingDirectory = "/path/to/models";
    };
  };
```

Then reproduce the demo like this:

```
curl http://localhost:5002/api/tts\\?text\\=Dear%20NixOS%20Maintainer,%20I%20am%20Mozilla%20TTS.%20I%20am%20ready%20to%20be%20merged,%20please%20review%20me\\. > /tmp/nixpkgs.wav
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/audio/tts/default.nix [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  preBuild = ''
+    # numba jit tries to write to its cache directory
+    export HOME=$TMPDIR
+    sed -i -e 's!tensorflow==.*!tensorflow!' requirements.txt
+    sed -i -e 's!librosa==[^"]*!librosa!' requirements.txt setup.py
+    sed -i -e 's!unidecode==[^"]*!unidecode!' requirements.txt setup.py
+    sed -i -e 's!bokeh==[^"]*!bokeh!' requirements.txt setup.py
+    sed -i -e 's!numba==[^"]*!numba!' requirements.txt setup.py
+    sed -i -e '/pylint/d' requirements.txt setup.py
+    sed -i -e '/cardboardlint/d' requirements.txt setup.py
```
Comment (Mic92) [inline pkgs/tools/audio/tts/default.nix:56]:
```suggestion
    # Not required for building/installation but for their development/ci workflow
    sed -i -e '/pylint/d' requirements.txt setup.py
    sed -i -e '/cardboardlint/d' requirements.txt setup.py
```

File: pkgs/tools/audio/tts/default.nix [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    tensorboardx
+    fuzzywuzzy
+    tensorflow_2
+    inflect
+    gdown
+    pysbd
+  ];
+
+  postInstall = ''
+    cp -r TTS/server/templates/ $out/${python3.sitePackages}/TTS/server
```
Comment (Mic92) [inline pkgs/tools/audio/tts/default.nix:81]: In future we could even try to fix upstream here and make their build system installing these templates.

File: pkgs/tools/audio/tts/default.nix [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+# Vocoder model:
+#   Multi-Band MelGAN
+#   https://drive.google.com/drive/folders/1XeRT0q4zm5gjERJqwmX5w84pMrD00cKD
+#
+# Start like this:
+#   tts-server \
+#     --vocoder_config /tmp/tts/vocoder/config.json \
+#     --vocoder_checkpoint /tmp/tts/vocoder/checkpoint_1450000.pth.tar \
+#     --tts_config /tmp/tts/config.json \
+#     --tts_checkpoint /tmp/tts/checkpoint_130000.pth.tar
```
Comment (Mic92) [inline pkgs/tools/audio/tts/default.nix:23]: The default configuration contains also a `scale.py` which is read from the current directory. So `cd /tmp/tts` is required for this example to work.

File: pkgs/tools/audio/tts/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/tools/audio/tts/default.nix:82]: Yeah, lots of stuff in their packaging could be improved.

File: pkgs/tools/audio/tts/default.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/tools/audio/tts/default.nix:39]: Expanded on the necessary directoy structure and the pwd.

File: pkgs/tools/audio/tts/default.nix [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+#   vocoder/config.json
+#   vocoder/scale_stats.npy
+#
+# Start like this:
+#   cd /tmp/tts
+#   tts-server \
+#     --vocoder_config /tmp/tts/vocoder/config.json \
+#     --vocoder_checkpoint /tmp/tts/vocoder/checkpoint_1450000.pth.tar \
+#     --tts_config /tmp/tts/config.json \
+#     --tts_checkpoint /tmp/tts/checkpoint_130000.pth.tar
```
Comment (Mic92) [inline pkgs/tools/audio/tts/default.nix:39]:
```suggestion
#     --vocoder_config ./vocoder/config.json \
#     --vocoder_checkpoint ./vocoder/checkpoint_1450000.pth.tar \
#     --tts_config ./config.json \
#     --tts_checkpoint ./checkpoint_130000.pth.tar
```

File: pkgs/tools/audio/tts/default.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/tools/audio/tts/default.nix:75]: How is this solved?

**Review by Mic92 — APPROVED** [review]
only nitpicks left.

**Comment by mweinelt** [comment]
All nitpicks addressed I believe.

**Review by Mic92 — APPROVED** [review]

**Comment by NickHu** [comment]
Perhaps off-topic, but I'd love to see rhasspy packaged - happy to help out too

**Comment by mweinelt** [comment]
Much of rhasspy is packaged over Mic92's nur-packages:
https://github.com/Mic92/nur-packages/commits/master

I'm currently running it like so:

```
{ config, pkgs, ... }:
let
  nur = import (builtins.fetchTarball "https://github.com/Mic92/nur-packages/archive/master.tar.gz") {
    inherit pkgs;
  };
in
{
  systemd.services.rhasspy = {
    after = [
      "network.target"
    ];
    wantedBy = [ "multi-user.target" ];
    environment = {
      XDG_RUNTIME_DIR = "/run/user/1000";
    };
    path = with pkgs; [
      picotts
      diffutils
      perl
    ];
    serviceConfig = {
      User = "hexa";
      Group = "users";
      WorkingDirectory = config.users.users.hexa.home;
    };
    script = ''
      ${nur.rhasspy}/bin/rhasspy --profile en | ${pkgs.utillinux}/bin/logger
    '';
  };
}
```

**Pull request merged**